tests/closures.test.ts

```diff
@@ -162,4 +162,155 @@ describe("Closures group's test cases on the proposal", () => {
   f(10)()
   `;
   assert("10. An escaping function calls its non-escaping sibling", src, PyInt(11));
+
+  src = `
+  def concat(items: [bool], stuff: [bool]) -> [bool]:
+    concatted : [bool] = None
+    concatted = items + stuff
+    return concatted
+
+  items : [bool] = None
+  stuff : [bool] = None
+  concatted : [bool] = None
+  items = [True, True, False]
+  stuff = [False, True]
+  concatted = concat(items, stuff)
+`;
+  asserts("11. A function with list created", [
+    [src, PyNone()],
+    ["concatted[0]", PyBool(true)],
+    ["concatted[1]", PyBool(true)],
+    ["concatted[2]", PyBool(false)],
+    ["concatted[3]", PyBool(false)],
+    ["concatted[4]", PyBool(true)],
+  ]);
+
+  src = `
+  def f(x: [int]) -> [int]:
+    def inc() -> [int]:
+      return x + [1]
+    return inc()
+  x : [int] = None
+  x = f([1])
+  x = f(x)
+  x = f(x)
+  `;
+  asserts("12. A nested function with list created", [
+    [src, PyNone()],
+    ["x[0]", PyInt(1)],
+    ["x[1]", PyInt(1)],
+    ["x[2]", PyInt(1)],
+  ]);
+
+  describe.skip("nested func - list", function () {
+    src = `
+  def f(x : [int]) -> [int]:
+    def g(y : [int]) -> [int]:
+      return x + h(y)
+    def h(z : [int]) -> [int]:
+      nonlocal x
+      x = z
+      return x + [1]
+    return g([10]) + g([7])
+
+  x: [int] = None
+  x = f([6])
+  `;
+    asserts("13. A nested function with `nonlocal` and list created", [
+      // [6, 10, 1, 10, 7, 1]
+      [src, PyNone()],
+      ["x[0]", PyInt(6)],
+      ["x[1]", PyInt(10)],
+      ["x[2]", PyInt(1)],
+      ["x[3]", PyInt(10)],
+      ["x[4]", PyInt(7)],
+      ["x[5]", PyInt(1)],
+    ]);
+  });
+
+  src = `
+  def f(x : str) -> str:
+    def g(y : str) -> str:
+      l:str = ""
+      l = h(y)
+      print(x)
+      return l
+    def h(z : str) -> str:
+      nonlocal x
+      print(x)
+      x = z
+      return x
+    return g("10")
+
+  x: str = ""
+  x = f("6")
+  print(x)
+`;
+  assertPrint("14. A nested function with `nonlocal` and string ", src, ["6", "10", "10"]);
+
+  src = `
+  def f(x: [int,int]) -> [int,int]:
+    def inc() -> [int, int]:
+      x.update({1:2})
+      return x
+    return inc()
+  x : [int, int] = None
+  x = {11:22, 21:44, (30+1):56, 4:55}
+  x = f(x)
+  `;
+  asserts("15. A nested function with dict", [
+    [src, PyNone()],
+    ["x[1]", PyInt(2)],
+    ["x[11]", PyInt(22)],
+    ["x[31]", PyInt(56)],
+  ]);
+
+  describe.skip("nested func - loop", function () {
```

Comment:
Another skipped case is about the `for` loop. We expect some big change to make it work as currently the assignee is a string rather than a `Destructure` , which is an obstacle for wrapping and pass around this variable. If we have time. we can work on this later with an additional PR after the final PR of the for-loop team got merged.